### PR TITLE
Performance-Optimzation: Do not recompile Pattern for each file/line

### PR DESCRIPTION
**Pull request by fabotronix**
The Singleline- and MultilineDetector were recompiling the regex pattern
for each checked file.

The compiled pattern can be cached to improve runtime performance.

### CODE REVIEW

**Comment by romani** [comment]
thanks a lot.

but build will fail, as we demand some format in commit message.

please update commit message with reference to this PR number.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/regexp/DetectorOptions.java [hunk 1]
```diff
@@ -69,6 +71,7 @@
      */
     public DetectorOptions setFormat(String format) {
         this.format = format;
+        pattern = null;
```
Comment (mkordas) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/regexp/DetectorOptions.java]: @fabotronix 
Can you explain why we need to assign `null` here? Doesn't it work without this line?

Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/regexp/DetectorOptions.java]: he need to reset pattern each time smb/smth call setters that are constituents of Pattern: ignore case option and format.

**Comment by romani** [comment]
@fabotronix , I am just curious .... 
please describe how you caught this case ? a tool that do analysis. 
You suffer from this that much to dig our code (embedding, .... ) ? ....

**Comment by fabotronix** [comment]
We are using checkstyle in a huge project (with many regex checkstyle rules) and a checkstyle run takes some time. So I did some sampling with jvisualvm, where I saw that it used more time to compile patterns than to actually check for them.

May be I am just a little impatient ;-)

**Comment by romani** [comment]
yes, we have bunch of places where optimization could be done. Right now we are focused on functional changes (false-positives, exceptions, ... ). 

I am postponing your fix till next release (end of November) as I need to review your change more attentively.

Can you share with us some numbers of how that fix improve performance, just be aware.

**Comment by fabotronix** [comment]
In our case the runtime for the checkstyle task nearly halved:

master with this patch:
 119437ms  :checkstyleMain
master without this patch:
 219262ms  :checkstyleMain
checkstyle 6.5:
 214020ms  :checkstyleMain

Of course your mileage may vary.

**Comment by romani** [comment]
merged as FF, thanks a lot!

more proper update will be done during #2523 .

**Pull request closed without merging**